File: WinForms/VB/HtmlMainDemo/Modules/Form/FormModule.vb
Imports System.Collections.Generic
Imports DevExpress.DXperience.Demos
Imports DevExpress.XtraEditors
Imports DevExpress.Utils.Html
Imports DevExpress.Utils.DirectXForm.Painter
Imports DevExpress.LookAndFeel
Imports System.Windows.Forms
Imports System
Imports DevExpress.Skins.XtraForm
Imports System.Drawing
Imports DevExpress.Utils
Imports DevExpress.Utils.DPI
Imports DevExpress.Utils.Drawing.Helpers

Namespace DevExpress.HTML.Demos

    Public Partial Class FormModule
        Inherits TutorialControlBase

        Private previewForm As PreviewDirectXForm = New PreviewDirectXForm()

        Public Sub New()
            InitializeComponent()
            Dim customTemplate = New HtmlTemplate(previewForm.HtmlTemplate)
            customTemplate.Tag = "Custom HTML-CSS Template"
            previewForm.Text = "DirectX Form - Default HTML-CSS Template"
            previewForm.Size = New Size(800, 600)
            previewForm.Clear()
            AddHandler previewForm.HtmlElementMouseClick, AddressOf Form_HtmlElementMouseClick
            Tutorials.TodoApp.PresentationHelper.CenterParent(hostPanel, previewForm)
            templateViewer.Init(previewForm, New HtmlTemplateCollection(New List(Of HtmlTemplate) From {customTemplate, defaultTemplate}))
            Dim previewContent = New HtmlContentControl()
            previewContent.Parent = previewForm
            previewContent.Dock = DockStyle.Fill
            previewContent.HtmlTemplate.Assign(contentTemplate)
            AddHandler previewContent.ElementMouseClick, Sub(s, e)
                If Equals(e.ElementId, "showform") Then ShowForm()
            End Sub
        End Sub

        Private Sub ShowForm()
            Using form = New DemoDirectXForm()
                form.HtmlTemplate.Assign(previewForm.HtmlTemplate)
                form.Size = ScaleDPI.ScaleSize(New Size(1000, 600))
                form.StartPosition = FormStartPosition.CenterScreen
                form.ShowDialog(Me)
            End Using
        End Sub

        Private Sub Form_HtmlElementMouseClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            If Equals(e.ElementId, "infobutton") Then
                BeginInvoke(New Action(Sub()
                    Using aboutForm = New AboutDirectXForm(aboutFormTemplate, svgImages, ScaleDPI)
                        aboutForm.ShowDialog(Me)
                    End Using
                End Sub))
            End If
        End Sub

#Region "Utils"
        Protected Overrides Sub DoShow()
            MyBase.DoShow()
            If previewForm IsNot Nothing AndAlso Not previewForm.Visible Then previewForm.Show()
        End Sub

        Protected Overrides Sub DoHide()
            If previewForm IsNot Nothing AndAlso previewForm.Visible Then previewForm.Hide()
            MyBase.DoHide()
        End Sub

        Private Sub OnPanelSizeChanged(ByVal sender As Object, ByVal e As EventArgs)
            If previewForm Is Nothing Then Return
            Dim size = CType(sender, Control).Size
            Dim padding As Integer = ScaleDPI.ScaleHorizontal(20)
            Dim minWidth As Integer = ScaleDPI.ScaleHorizontal(600)
            Dim minHeight As Integer = ScaleDPI.ScaleHorizontal(800)
            Dim maxSize As Size = ScaleDPI.ScaleSize(New Size(1000, 750))
            Dim height As Integer = Math.Min(size.Height - padding * 2, maxSize.Height)
            Dim width As Integer = Math.Min(size.Width - padding * 2, maxSize.Width)
            If height > width * 3 \ 4 Then
                height = width * 3 \ 4
            Else
                width = height * 4 \ 3
            End If

            If width < minWidth Then width = minWidth
            If height < minHeight Then height = minHeight
            If height > width * 3 \ 4 Then
                height = width * 3 \ 4
            Else
                width = height * 4 \ 3
            End If

            previewForm.SetBounds((size.Width - width) \ 2, (size.Height - height) \ 2, width, height)
        End Sub
#End Region  ' Utils
    End Class

#Region "PreviewForm"
    Friend Class PreviewDirectXForm
        Inherits DemoDirectXForm

        Protected Overrides Function CreateFormBorderPainter() As FormPainter
            Return New PreviewDirectXFormPainter(Me, LookAndFeel)
        End Function

        Friend Sub Clear()
            For Each control As Control In Controls
                If control IsNot ContainerControl Then control.Dispose()
            Next
        End Sub
    End Class

    Friend Class PreviewDirectXFormPainter
        Inherits DirectXFormPainter

        Public Sub New(ByVal owner As DirectXForm, ByVal lf As UserLookAndFeel)
            MyBase.New(owner, lf)
        End Sub

        Protected Overrides Sub WMNCHitTest(ByRef msg As Message)
            MyBase.WMNCHitTest(msg)
            If msg.Result <> IntPtr.Zero Then msg.Result = New IntPtr(Utils.Drawing.Helpers.NativeMethods.HT.HTCLIENT)
        End Sub

        Protected Overrides Sub OnClick(ByVal e As MouseEventArgs, ByVal kind As FormCaptionButtonAction)
        End Sub
    End Class

#End Region  ' PreviewForm
    Public Class AboutDirectXForm
        Inherits DirectXForm

        Public Sub New(ByVal template As HtmlTemplate, ByVal htmlImages As Object, ByVal scaleDpi As ScaleHelper)
            MyBase.New()
            Me.HtmlImages = htmlImages
            HtmlTemplate.Assign(template)
            StartPosition = FormStartPosition.CenterParent
            Size = scaleDpi.ScaleSize(FormDefaultSize)
            FormBorderStyle = FormBorderStyle.FixedDialog
            AddHandler HtmlElementMouseClick, AddressOf OnHtmlElementMouseClick
        End Sub

        Private Shared ReadOnly FormDefaultSize As Size = New Size(660, 460)

        Protected Overrides Sub WndProc(ByRef m As Message)
            MyBase.WndProc(m)
            If m.Msg = MSG.WM_DPICHANGED Then Size = ScaleDPI.ScaleSize(FormDefaultSize)
        End Sub

        Private Sub OnHtmlElementMouseClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            If e.Element Is Nothing Then Return
            Dim id = e.Element.Id
            Dim parentId = e.Element.ParentElement?.Id
            If Equals(id, "contentbutton") OrElse Equals(parentId, "contentbutton") Then
                Close()
            End If
        End Sub
    End Class
End Namespace

File: WinForms/CS/HtmlMainDemo/Modules/Form/FormModule.cs
using System.Collections.Generic;
using DevExpress.DXperience.Demos;
using DevExpress.XtraEditors;
using DevExpress.Utils.Html;
using DevExpress.Utils.DirectXForm.Painter;
using DevExpress.LookAndFeel;
using System.Windows.Forms;
using System;
using DevExpress.Skins.XtraForm;
using System.Drawing;
using DevExpress.Utils;
using DevExpress.Utils.DPI;
using DevExpress.Utils.Drawing.Helpers;

namespace DevExpress.HTML.Demos {
    public partial class FormModule : TutorialControlBase {
        PreviewDirectXForm previewForm = new PreviewDirectXForm();

        public FormModule() {
            InitializeComponent();
            var customTemplate = new HtmlTemplate(previewForm.HtmlTemplate);
            customTemplate.Tag = "Custom HTML-CSS Template";

            previewForm.Text = "DirectX Form - Default HTML-CSS Template";
            previewForm.Size = new Size(800, 600);
            previewForm.Clear();
            previewForm.HtmlElementMouseClick += Form_HtmlElementMouseClick;
            DevExpress.Tutorials.TodoApp.PresentationHelper.CenterParent(hostPanel, previewForm);

            templateViewer.Init(previewForm, new HtmlTemplateCollection(new List<HtmlTemplate> {
                customTemplate,
                defaultTemplate,
            }));

            var previewContent = new HtmlContentControl();
            previewContent.Parent = previewForm;
            previewContent.Dock = DockStyle.Fill;
            previewContent.HtmlTemplate.Assign(contentTemplate);
            previewContent.ElementMouseClick += (s, e) => { 
                if(e.ElementId == "showform") ShowForm(); 
            };
        }

        void ShowForm() {
            using(var form = new DemoDirectXForm()) {
                form.HtmlTemplate.Assign(previewForm.HtmlTemplate);
                form.Size = ScaleDPI.ScaleSize(new Size(1000, 600));
                form.StartPosition = FormStartPosition.CenterScreen;
                form.ShowDialog(this);
            }
        }

        void Form_HtmlElementMouseClick(object sender, DxHtmlElementMouseEventArgs e) {
            if(e.ElementId == "infobutton") {
                BeginInvoke(new Action(() => {
                    using(var aboutForm = new AboutDirectXForm(aboutFormTemplate, svgImages, ScaleDPI)) {
                        aboutForm.ShowDialog(this);
                    }
                }));
            }
        }

        #region Utils
        protected override void DoShow() {
            base.DoShow();
            if(previewForm != null && !previewForm.Visible)
                previewForm.Show();
        }
        protected override void DoHide() {
            if(previewForm != null && previewForm.Visible)
                previewForm.Hide();
            base.DoHide();
        }

        void OnPanelSizeChanged(object sender, EventArgs e) {
            if(previewForm == null)
                return;
            var size = ((Control)sender).Size;
            int padding = ScaleDPI.ScaleHorizontal(20);
            int minWidth = ScaleDPI.ScaleHorizontal(600);
            int minHeight = ScaleDPI.ScaleHorizontal(800);
            Size maxSize = ScaleDPI.ScaleSize(new Size(1000, 750));
            int height = Math.Min(size.Height - padding * 2, maxSize.Height);
            int width = Math.Min(size.Width - padding * 2, maxSize.Width);
            if(height > width * 3 / 4)
                height = width * 3 / 4;
            else width = height * 4 / 3;
            if(width < minWidth)
                width = minWidth;
            if(height < minHeight)
                height = minHeight;
            if(height > width * 3 / 4)
                height = width * 3 / 4;
            else width = height * 4 / 3;
            previewForm.SetBounds((size.Width - width) / 2, (size.Height - height) / 2, width, height);
        }
        #endregion Utils
    }

    #region PreviewForm
    class PreviewDirectXForm : DemoDirectXForm {
        protected override FormPainter CreateFormBorderPainter() {
            return new PreviewDirectXFormPainter(this, LookAndFeel);
        }
        internal void Clear() {
            foreach(Control control in Controls) {
                if(control != ContainerControl)
                    control.Dispose();
            }
        }
    }
    class PreviewDirectXFormPainter : DirectXFormPainter {
        public PreviewDirectXFormPainter(DirectXForm owner, UserLookAndFeel lf) : base(owner, lf) { }
        protected override void WMNCHitTest(ref Message msg) {
            base.WMNCHitTest(ref msg);
            if(msg.Result != IntPtr.Zero)
                msg.Result = new IntPtr(Utils.Drawing.Helpers.NativeMethods.HT.HTCLIENT);
        }
        protected override void OnClick(MouseEventArgs e, FormCaptionButtonAction kind) { }
    }
    #endregion PreviewForm

    public class AboutDirectXForm : DirectXForm {
        public AboutDirectXForm(HtmlTemplate template, object htmlImages, ScaleHelper scaleDpi) : base() {
            HtmlImages = htmlImages;
            HtmlTemplate.Assign(template);
            StartPosition = FormStartPosition.CenterParent;
            Size = scaleDpi.ScaleSize(FormDefaultSize);
            FormBorderStyle = FormBorderStyle.FixedDialog;
            HtmlElementMouseClick += OnHtmlElementMouseClick;
        }
        static readonly Size FormDefaultSize = new Size(660, 460);
        protected override void WndProc(ref Message m) {
            base.WndProc(ref m);
            if(m.Msg == MSG.WM_DPICHANGED)
                Size = ScaleDPI.ScaleSize(FormDefaultSize);
        }
        void OnHtmlElementMouseClick(object sender, DxHtmlElementMouseEventArgs e) {
            if(e.Element == null)
                return;
            var id = e.Element.Id;
            var parentId = e.Element.ParentElement?.Id;
            if(id == "contentbutton" || parentId == "contentbutton") {
                Close();
            }
        }
    }
}
